### PR TITLE
[Filesystem] HTTP Directory parser enhancements for multi-line anchor tags

### LINKED ISSUE #17623: Failed parsing some file links in html index file when using "web server directory" protocol
### Describe the bug

When using http or https ("web server directory" in Kodi) protocol to watch video, the file list is a html file containing hyper-links of files, fetched from the web server.

I found that some valid html links in index files can't be parsed/displayed properly, the video files won't show up in the file list.

## Expected Behavior

I expect that every file links in the html file with valid html syntax will be properly parsed and displayed in the file list

## Actual Behavior

The files aren't displayed in the file list

## Possible Fix

Since I didn't trace the source code so I don't have strong idea here, but maybe need a stronger html parser

### To Reproduce

1. Setup an http server with a `index.html` text file use following content:

```html
<html>
  <head>
    <title>Page title</title>
  </head>
  <body>
    <a href="video.mkv">
      video.mkv
    </a>
    <a href="video0.mkv">video0.mkv</a> <a href="video1.mkv">video1.mkv</a>
    <a href="video2.mkv">
      video2.mkv
    </a>
    <a href="video3.mkv">video3.mkv</a>
  </body>
</html>
```

The html tag `<a> </a>` of file `video3.mkv` is in the same single line, and the `video.mkv` `video2.mkv` are using multi-line separately, `video0.mkv` `video1.mkv` shares the same line.

2. Add a Video source in Kodi, using web server directory protocol, and navigate to the http server that contains the testing `index.html` from previous step

3. Visit the added file source to see the file list, you'll see that only `video3.mkv` is displayed, and the all the others are not.

### Debuglog

The debuglog can be found here, I have the same issue in all my Kodi environment, so I only attach one log from one of my laptops that using Linux Mint, hope it should be enough:

[kodi.log](https://github.com/xbmc/xbmc/files/4442684/kodi.log)

## Additional context

Some http servers like [Caddy](https://caddyserver.com/) generate file list index in multiple line format, which is valid html, but can't be used by Kodi.

### Screenshots 
Here is the screenshots to help explain the problem:

Screenshot of the bug reproduce procedure:
image

Screenshot of the file source from Caddy server generated file list:
image

### My Environment

Used Operating system(though I thought that the problem is OS independent):

 - [x] Ubuntu 18.04
 - [x] Linux Mint 19.3
 - [x] Raspberry-Pi 3 B+ With LibreELEC 9.2.1 / OSMC 2020.03-1
 - [x] Windows 10
 - Kodi version: v18.6 Leia

### PR DESCRIPTION
**Pull request by djp952**
## Description
This PR attempts to address [Issue 17623](https://github.com/xbmc/xbmc/issues/17623).  I was poking around the Issues and thought I might be able to help with this one.

## Motivation and Context
The referenced Issue indicates that generic HTML that may include multiple anchor tags on one line, or multi-line anchor tags do not work with "Web server directory" network locations.  The current implementation reads and parses the source data on a line-by-line basis, making multi-line tags impossible, and had a limitation with the regular expression that prevented capture of multiple anchors on a single line.

The proposed implementation processes the entire index HTML file rather than one line at a time and adjusts the regular expressions accordingly.  The primary regular expression now captures both the anchor tag and any metadata (date/time, size) that follows it, until a new anchor tag or the end of the data has been detected.  Adjustments needed to be made to the date/time and size capture expressions as well to accommodate.

Note that Internet Information Server (IIS) format is still not supported, as the metadata prefixes the anchor tags in that format.

I strongly suggest a thorough review by one or more devs that use the feature a lot to make sure I didn't cause any regressions or if the updated regular expression(s) are inadequate.

## How Has This Been Tested?
This was tested on Windows x64 using local build of master branch from 4/9/2020.  Testing of the updated regular expressions and data capture was performed using the following web server instances, also on Windows x64:

- Apache 2.4.43 with bare index format
- Apache 2.4.43 with fancy index format
- Apache 2.4.43 with HTML table index format
- Lighttp 1.4.49.1 with default index format
- Nginx 1.16.1 with default index format
- Raw HTML provided with [Issue 17623](https://github.com/xbmc/xbmc/issues/17623)

The list of server types to test was generated by searching the Kodi forums as well as using the information from the code itself. I doubt this list will be exhaustive.

I would appreciate any links to public sites that I can also test these changes against.  For each server type, a typical step-through/code coverage test method was used to verify the captures were as expected.

## Screenshots (if appropriate):

Contents of parsed HTTP Directory using HTML provided in Issue 17623:

issue-view

Contents of parsed HTTP Directory using nginx server, other servers show same data with exception of Apache bare format since it does not provide date/time or size information:

nginx-size
nginx-date

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

**Comment by djp952** [comment]
Updated to remove an unnecessary local variable that was used for debugging purposes only.

**Comment by djp952** [comment]
Updated to correct build error on non-Windows targets (I passed an r-value into StringUtils::Trim)

**Comment by lrusak** [comment]
Might be worth writing some test units to test various index.html layouts? You could use our build in webserver to serve the files.

See: https://github.com/xbmc/xbmc/blob/master/xbmc/network/test/TestWebServer.cpp

**Comment by djp952** [comment]
Definitely! Would you mind throwing a "Don't Merge" tag on this to make sure it doesn't move forward without the tests? It will take me some evenings to work on that. Thanks!!

**Comment by djp952** [comment]
Hi @lrusak, I managed to figure out to add the unit tests and have added them.  It was definitely a good idea as there was still a defect in one of the regular expressions.  If you think the unit test seems OK, can you remove the WIP tag so it will run through Jenkins?  Thank you as always!

File: xbmc/filesystem/HTTPDirectory.cpp [hunk 1]
```diff
@@ -42,31 +41,33 @@ bool CHTTPDirectory::GetDirectory(const CURL& url, CFileItemList &items)
   }
 
   CRegExp reItem(true); // HTML is case-insensitive
-  reItem.RegComp("<a href=\"(.*)\">(.*)</a>");
+  reItem.RegComp("<a href=\"(.*?)\">\\s*(.*?)\\s*</a>(.+?)(?=<a|</tr|$)");
 
-  CRegExp reDateTime(true);
-  reDateTime.RegComp("<td align=\"right\">([0-9]{2})-([A-Z]{3})-([0-9]{4}) ([0-9]{2}):([0-9]{2}) +</td>");
+  CRegExp reDateTimeHtml(true);
+  reDateTimeHtml.RegComp("<td align=\"right\">([0-9]{2})-([A-Z]{3})-([0-9]{4}) ([0-9]{2}):([0-9]{2}) +</td>");
```
Comment (Rechi) [inline xbmc/filesystem/HTTPDirectory.cpp:47]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

File: xbmc/filesystem/HTTPDirectory.cpp (hunk 1 above)
Comment (djp952) [reply xbmc/filesystem/HTTPDirectory.cpp]: sorry, fixed

**Comment by phunkyfish** [comment]
Nice work. Are we happy with this with the tests or should it also go in a test build?

**Comment by djp952** [comment]
Unless there are any other specific web server(s) to add unit test cases for, I'm pretty confident it handles any formats that it previously handled.  Not sure if that helps any decisions, of course.  My knowledge of what web servers folks may actually be using is anecdotal at best.

**Comment by phunkyfish** [comment]
@djp952 do you feel that this would benefit from some user testing before merging?

If you think it could use some then we can ask @MilhouseVH to include it in a test build.

**Comment by djp952** [comment]
Absolutely.  I don't use this feature of the application myself, would greatly prefer someone that actually uses it has a chance to ensure it doesn't introduce any problems.

**Comment by phunkyfish** [comment]
Cool, @MilhouseVH can you add this to a test build? Ty 😉

**Comment by MilhouseVH** [comment]
I'll include this starting with `#0422`.

**Comment by phunkyfish** [comment]
@djp952 generally with a change like this we add to the test libreelec builds for a few days and if we don't get any negative feedback then we merge. Cool?

**Comment by djp952** [comment]
Of course! Thanks!

**Comment by phunkyfish** [comment]
@MilhouseVH any feedback?

**Comment by phunkyfish** [comment]
@djp952 would you like to join kodi slack? If so PM me your email address on the kodi forum.

**Comment by MilhouseVH** [comment]
> @MilhouseVH any feedback

No feedback, however it's possible nobody has really tested this change as it's a little bit niche.

**Comment by phunkyfish** [comment]
Let's give it another few days so.

**Comment by phunkyfish** [comment]
No feedback on this so will merge later today, thanks @djp952
